vault-plugins/flant_iam_auth/usecase/authn/accesstoken/validator.go

```diff
@@ -0,0 +1,319 @@
+	"github.com/hashicorp/go-cleanhttp"
+	"golang.org/x/oauth2"
+	"gopkg.in/square/go-jose.v2/jwt"
+
+	"github.com/flant/negentropy/vault-plugins/flant_iam_auth/model"
+	jwt2 "github.com/flant/negentropy/vault-plugins/flant_iam_auth/usecase/authn/jwt"
+)
+
+// DefaultLeewaySeconds defines the amount of leeway that's used by default
+// for validating the "nbf" (Not Before) and "exp" (Expiration Time) claims.
+const DefaultLeewaySeconds = 150
+
+type accessTokenValidator struct {
+	userinfoURL string
+	issuer      string
+	client      *http.Client
+}
+
+func NewAccessTokenValidator(ctx context.Context, cfg *model.AuthSource) (jwt2.JwtValidator, error) {
+	if cfg == nil || cfg.OIDCDiscoveryURL == "" {
+		return nil, fmt.Errorf("%v", cfg)
+	}
+	caCtx, err := createCAContext(ctx, cfg.OIDCDiscoveryCAPEM)
+	if err != nil {
+		return nil, err
+	}
+	client := http.DefaultClient
+	if c, ok := caCtx.Value(oauth2.HTTPClient).(*http.Client); ok {
+		client = c
+	}
+
+	wellKnown := strings.TrimSuffix(cfg.OIDCDiscoveryURL, "/") + "/.well-known/openid-configuration"
+	req, err := http.NewRequest(http.MethodGet, wellKnown, nil)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := client.Do(req.WithContext(caCtx))
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	// Read the response body and status code
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read response body: %v", err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("%s: %s", resp.Status, body)
+	}
+
+	// Unmarshal the response body to obtain the issuer and JWKS URL
+	var p struct {
+		Issuer           string `json:"issuer"`
+		UserInfoEndpoint string `json:"userinfo_endpoint"`
+	}
+	if err := unmarshalResp(resp, body, &p); err != nil {
+		return nil, fmt.Errorf("failed to decode OIDC discovery document: %w", err)
+	}
+
+	return &accessTokenValidator{
+		userinfoURL: p.UserInfoEndpoint,
+		issuer:      p.Issuer,
+		client:      client,
+	}, nil
+}
+
+// Validate validates access_token JWT
+//
+// The given JWT is considered valid if:
+//  0. iss match issuer at OIDCDiscoveryURL
+//  1. oidc-provider returns userinfo presenting access_token
+//  2. Its claims set, header parameter and userinfo response values match what's given by Expected.
+//  3. It's valid with respect to the current time. This means that the current
+//     time must be within the times (inclusive) given by the "nbf" (Not Before)
+//     and "exp" (Expiration Time) claims and after the time given by the "iat"
+//     (Issued At) claim, with configurable leeway. See Expected.Now() for details
+//     on how the current time is provided for validation.
+func (v *accessTokenValidator) Validate(ctx context.Context, token string, expected hcjwt.Expected) (map[string]interface{}, error) {
+	// First, verify the signature to ensure subsequent validation is against verified claims
+	allClaims := map[string]interface{}{}
```

Comment:
why do create empty map, enrichedClaims returns it and rewrite anyway

Reply:
+

Reply:
done